realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -281,6 +342,41 @@ public void refresh() {
     }
 
     /**
+     * Blocks the current thread util transactions are committed by other threads to this Realm or
+     * {@link #stopWaitForChange()} gets called.
+     *
+     * @return {@code true} if transactions are committed to this Realm, {@code false} if
+     * {@link #stopWaitForChange()} gets called.
+     * @throws IllegalStateException if attempting to wait within a transaction.
+     */
+    public boolean waitForChange() {
+        checkIfValid();
+        if (isInTransaction()) {
+            throw new IllegalStateException("Cannot wait for changes inside of a transaction.");
+        }
+        sharedGroupManager.setWaitForChangeEnabled(true);
+        // add this to wait list for batch stop
+        addRealmToWaitList(this);
+        boolean hasChanged = sharedGroupManager.waitForChange();
+        // remove this from wait list
+        removeRealmFromWaitList(this);
+        if (hasChanged) {
+            sharedGroupManager.advanceRead();
```

Comment:
@beeender `HandlerController.updateAsyncQueries()` _could_ update all change listeners. If we're to let it happen, we can simply pass a `REALM_CHANGED` to a handler and all is well. I think `waitForChange` should notify change listeners.
```java
@Test
public void waitForChange_onLooperThread() throws InterruptedException {
    final long numberOfBlocks = 16;
    final long numberOfObjects = 1; // must be greater than 1
    final CountDownLatch openedLatch = new CountDownLatch(1);
    final CountDownLatch changedLatch = new CountDownLatch(1);
    final CountDownLatch closedLatch = new CountDownLatch(2);
    final AtomicBoolean result1 = new AtomicBoolean(false);
    final AtomicBoolean result2 = new AtomicBoolean(false);
    final AtomicLong count1 = new AtomicLong(0);
    final AtomicLong count2 = new AtomicLong(0);

    Thread thread = new Thread(new Runnable() {
        @Override
        public void run() {
            Looper.prepare();
            // create background realm
            final Realm realm = Realm.getInstance(realmConfig);
            openedLatch.countDown();
            final Runnable endTest = new Runnable() {
                @Override
                public void run() {
                    closedLatch.countDown();
                    if (closedLatch.getCount() == 0) {
                        realm.close();
                    }
                }
            };
            // first async results
            Realm.asyncQueryExecutor.pause();
            final RealmResults<AllTypes> asyncResults1 = realm.where(AllTypes.class).findAllAsync();
            Realm.asyncQueryExecutor.resume();
            asyncResults1.addChangeListener(new RealmChangeListener() {
                @Override
                public void onChange() {
                    count1.set(asyncResults1.size());
                    endTest.run();
                }
            });
            result1.set(realm.waitForChange());
            changedLatch.countDown();
            // second async results
            Realm.asyncQueryExecutor.pause();
            final RealmResults<AnnotationIndexTypes> asyncResults2 = realm.where(AnnotationIndexTypes.class).findAllAsync();
            Realm.asyncQueryExecutor.resume();
            asyncResults2.addChangeListener(new RealmChangeListener() {
                @Override
                public void onChange() {
                    count2.set(asyncResults2.size());
                    endTest.run();
                }
            });
            result2.set(realm.waitForChange());
            Looper.loop();
        }
    });
    thread.start();

    TestHelper.awaitOrFail(openedLatch);
    realm.beginTransaction();
    realm.createObject(AllTypes.class);
    realm.commitTransaction();
    TestHelper.awaitOrFail(changedLatch);
    assertTrue(result1.get());
    populateForDistinct(realm, numberOfBlocks, numberOfObjects, false);
    TestHelper.awaitOrFail(closedLatch);
    assertTrue(result2.get());
    assertEquals(1, count1.get());
    assertEquals(numberOfBlocks * numberOfBlocks, count2.get());
}
```